lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,214 @@
+    PerformanceType,
+)
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchHttpError
+from charms.opensearch.v0.opensearch_internal_data import Scope
+from ops.framework import EventBase, EventSource
+
+# The unique Charmhub library identifier, never change it
+LIBID = "8b7aa39016e748ea908787df1d7fb089"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ApplyProfileTemplatesOpenSearch(EventBase):
+    """Attempt to acquire lock & restart OpenSearch.
+
+    The main reason to have a separate event, is to be able to wait for the cluster.
+    It defers otherwise and only defers the execution of this particular task.
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+
+        self.framework.observe(
+            charm.on.install,
+            self._on_install,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    def _on_install(self, event: EventBase):
+        """Handle install event.
+
+        Execute it on install, as defined in the docs, it is the first hook to run:
+           https://juju.is/docs/sdk/config-changed-event
+
+        install -> config-changed -> start
+
+        Store the current perf. profile we are applying. This must happen solely
+        on the beginning of the charm lifecycle. Therefore, we check if the unit is down.
+        """
+        if not os.path.exists(
+            os.path.join(
+                self.charm.opensearch.paths.conf,
+                self.charm.opensearch_config.JVM_OPTIONS,
+            )
+        ):
+            event.defer()
+            return
+        self.apply(self.charm.config.get(PERFORMANCE_PROFILE))
+
+    def _on_peer_cluster_relation_changed(self, _: EventBase):
+        """Handle the peer cluster relation changed event."""
+        if not (deployment_desc := self.charm.opensearch_peer_cm.deployment_desc()):
+            # Deployment description not available yet
+            # Nothing to do
+            return
+
+        if self.apply(deployment_desc.profile):
+            # We need to restart
+            self.charm.restart_event.emit()
+
+    @property
+    def current(self) -> OpenSearchPerfProfile | None:
```

Comment:
The `deployment_desc` does not get updated if we have small deployments. Reason is because I am setting the config as the following:
```
        if deployment_desc.typ == DeploymentType.MAIN_ORCHESTRATOR:
            # Only the main orchestrator can change the performance profile
            perf_profile_needs_restart = self.performance_profile.apply(
                self.config.get(PERFORMANCE_PROFILE)
            )
            if self.opensearch_peer_cm.is_provider(typ="main") and perf_profile_needs_restart:
                # Update the peer-cluster-orchestrator as needed
                self.peer_cluster_provider.refresh_relation_data(event) ## <<< happens if we have at least one peer-cluster-orch. requirer
```

Reply:
As explained above. The Deployment description object is unrelated to large deployments. :)

What is preventing us from moving the field from `deployment_desc.profile` to `deployment_desc.config.profile`?

Doing so will update the deployment_desc automatically on config changes.

Reply:
To clarify: I moved the profile entry to the `deployment_desc.config.profile`.

Now, that is not enough for this logic. As I have two answer basic one question:
`the state I am == the state I must be` >> this will trigger the profile change or not

Now, the question:
`the state I am` is defined by my **peer** relation databag. I store there the current status every time I update it

The question:
`the state I must be` is defined by either a config change (if this is a MAIN app) or by the `PeerClusterRelData.DeploymentDescription.config.profile`

So, I need the current method to know the profile.

Reply:
the state I am == can be based on the content of the opensearch.yml heap no?

Reply:
no, based on the `jvm.options`, the current jvm args (meaning, the process args for Xms and Xmx) and the index templates